service/history/workflowTaskHandler.go

```diff
@@ -1125,6 +1169,126 @@ func (handler *workflowTaskHandlerImpl) handleCommandModifyWorkflowProperties(
 	return err
 }
 
+func (handler *workflowTaskHandlerImpl) handleMessageAcceptWorkflowExecutionUpdate(
+	_ context.Context,
+	protocolInstanceID string,
+	updAcceptance *updatepb.Acceptance,
+) error {
+
+	handler.metricsHandler.Counter(metrics.MessageTypeAcceptWorkflowExecutionUpdateCounter.GetMetricName()).Record(1)
+
+	executionInfo := handler.mutableState.GetExecutionInfo()
+	nsID := namespace.ID(executionInfo.GetNamespaceId())
+
+	if err := handler.sizeLimitChecker.checkIfPayloadSizeExceedsLimit(
+		// TODO (alex-update): Should use MessageTypeTag here but then it needs to be another metric name too.
+		metrics.CommandTypeTag("MessageAcceptUpdate"),
+		updAcceptance.Size(),
+		"Message body of type update.Acceptance exceeds size limit.",
+	); err != nil {
+		return handler.failWorkflow(enumspb.WORKFLOW_TASK_FAILED_CAUSE_BAD_UPDATE_WORKFLOW_EXECUTION_MESSAGE, err)
+	}
+
+	if err := handler.validateCommandAttr(
+		func() (enumspb.WorkflowTaskFailedCause, error) {
+			return handler.attrValidator.validateWorkflowExecutionUpdateAcceptanceMessage(
+				nsID,
+				protocolInstanceID,
+				updAcceptance,
+			)
+		},
+	); err != nil || handler.stopProcessing {
+		return err
+	}
+
+	_, err := handler.mutableState.AddWorkflowExecutionUpdateAcceptedEvent(protocolInstanceID, updAcceptance)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (handler *workflowTaskHandlerImpl) handleMessageCompleteWorkflowExecutionUpdate(
+	_ context.Context,
+	protocolInstanceID string,
+	updResponse *updatepb.Response,
+) error {
+
+	handler.metricsHandler.Counter(metrics.MessageTypeCompleteWorkflowExecutionUpdateCounter.GetMetricName()).Record(1)
+
+	executionInfo := handler.mutableState.GetExecutionInfo()
+	nsID := namespace.ID(executionInfo.GetNamespaceId())
+
+	if err := handler.sizeLimitChecker.checkIfPayloadSizeExceedsLimit(
+		// TODO (alex-update): Should use MessageTypeTag here, but then it needs to be another metric name too.
+		metrics.CommandTypeTag("MessageCompleteUpdate"),
+		updResponse.Size(),
+		"Message body of type update.Response exceeds size limit.",
+	); err != nil {
+		return handler.failWorkflow(enumspb.WORKFLOW_TASK_FAILED_CAUSE_BAD_UPDATE_WORKFLOW_EXECUTION_MESSAGE, err)
```

Comment:
Will this add a workflow task failed event to the history for speculative WT?

Reply:
Yes, this seems to be broken and will lead to WTFailed event w/o preceding WTScheduled/WTStarted events. I will address this in separate PR.

Reply:
This was addressed here. Current behaviour: first speculative WT failure is not written to the history, but new WTis recreated as "normal", and got all corresponding events in the history, and if it keeps failing, it will become transient. I need to think if this should be changed in future.